README.md
# SerialPacker

This is an Arduino and Python library for sending and receiving
variable-sized, CRC16-protected data packets.

It is optimized for daisy-chaining multiple modules onto an ad-hoc bus,
i.e. each member of the chain forwards the data it receives to the next
device as quickly as possible, i.e. without waiting for the end of the
packet and without requiring RAM for the whole thing. This reduces RAM
usage and speeds up processing, particularly if your string is long.

Originally based on code by Stuart Pittaway.

## Usage

### Python

As the Python code is intended for the "master" node, it does not use
callbacks and does not support assembling or forwarding messages on the
fly.

The module exports a `SerialPacker` class which you can instantiate with
the following arguments:

* max\_idle: Milliseconds between intra-packet bytes. Default: 10 ms.

* crc: CRC generator class. Default: CRC16/0xBAAD.

* max\_packet: maximum packet length. Default: 127.

* frame\_start: frame start byte. The default is 0x85. `None` skips
  framing entirely.

#### CRC

The CRC class must be callable with an empty initializer.

Objects have the following attributes:

* feed(byte): add a byte to the CRC.

* crc: the accumulated CRC.

* len: the number of bytes to transmit.

#### Sending

The method `frame(data)` returns a head/tail tuple. Transmit the head
bytes, the data, and the tail bytes.

#### Receiving

Send all incoming bytes to `feed(byte)`. This method returns a complete
message when it is complete and its CRC matches. Otherwise, if the byte is
not part of a message, it is returned as-is.

#### Console data

If you set `frame_start` to a non-`None` value, incoming non-frame data
are accumulated in an internal buffer. You should periodically call
`read()` to zero and return it.

If the buffer ends with an incomplete UTF-8 character, it will not be
returned until `max_idle` milliseconds have passed since the last call to
`feed`.

### Arduino / C++

Define `SP_FRAME_START` to a byte that shall introduce a new packet.
Until that byte is seen, any other data will be ignored.
The default value is 0x85 (contains some alternating bits, is not a valid
ASCII or UTF-8 character). If `SP_FRAME_START` is defined to `-1`, any
character may start a message.

Define `SP_MAX_PACKET` as the maximum possible packet size. You do not need to
reserve space for the whole packet. If this value is >255, packet lengths
greater than 127 are encoded in two bytes. The default is 127 for maximum
compatibility.

Define `SP_MAX_FRAME_DELAY` as the number of milliseconds between successive
serial characters. A delay longer than this indicates the start of a new
frame. The default is 100.

Define `SP_SENDLEN` if you want to track the length of the transmitted
data. This adds code to pad a transmitted message automatically when you
abort due to an error, and prevents you from adding more data than you
should.

Define `SP_MARK` to a byte that's prefixed to every transmitted byte.
Also, received bytes are only considered part of a packet when they're
prefixed with that byte. This is useful if some embedded thing needs to
print debug information *while* sending a packet. We recommend `0x10`
("DLE").

Define `SP_NONFRAME_STREAM` if you want incoming characters that are not
part of a frame to be forwarded to another port. Obviously, this does not
work if `SP_FRAME_START` is not defined.

Define `SP_TRACE` to a serial stream (like `Serial1`) if you want to log a
mostly-readable hex copy of the data you're receiving.

Define `SP_CRC` to some suitable 16-bit polynomial. The default is 0xBAAD.
If you do use a different polynomial you need to patch the source to
include the appropriate lookup table.

Define `SP_ERRCOUNT` if you want to count receiver errors.

Include `<SerialPacker.h>`.

Create a SerialPacker instance, called SP in the rest of this document.

Set up your serial (or other) data stream.

Call `SP.begin(stream, header_handler, read_handler, packet_handler,
recv_buffer, recv_size, header_size)`, where

* `stream` is your data stream (obviously),

* `header_handler` is a function that's called when `header_size` bytes
  have been read.

* `read_handler` is a function that's called after additional bytes
  have been read after your header handler calls `sendDefer`.

* `packet_handler` is a function that's called when a packet has been
  received (with correct CRC),

* `recv_buffer` is a buffer of size `recv_size`. Bytes beyond this size
  are not stored, but possibly forwarded.

Periodically call `SP.checkInputStream()`.

#### Sending packet data

Call `sendStartFrame(SB_SIZE_T length)` to start transmitting a
packet.

Call `sendByte(uint8_t data)` to send a single byte, or `sendBuffer(const
uint8_t *buffer, SB_SIZE_T length)` to send multiple bytes.

Call `sendEndFrame(bool broken=false)` to transmit the CRC. If `broken` is
set, the CRC is intentionally mangled so that the next receiver will not
treat the packet as valid.

Sending more than the indicated number of bytes is not possible; they are
silently ignored.

#### Receiving packet data

Periodically call `SP.checkInputStream()`.

Your `onPacket` handler is called when a message is complete. If it is
longer than `bufferSize`, data exceeding this buffer have been discarded.
If you called `sendCopy` earlier, you need to call `sendEndFrame()` here.

#### Replying / Forwarding packet data

From within your `onHeader` handler, call `sendCopy(addLength)`. This sends
the header and the rest of the message onwards.

Alternately, call `sendDefer(readLength)`. This reads an additional
`readLength` bytes into the buffer without forwarding them, then calls your
`onRead` hook. You then call `sendCopy` (or another `sendDefer` if it's a
variable-length message) from there.

When the packet is complete, your `onPacket` handler *must* send exactly
`addLength` bytes.

If you decide that the frame should be invalidated, send filler bytes to
fulfill your `addLength` promise, then call `sendEndFrame(true)`. This
transmits an invalid CRC. If you do not do this 

### Error handling

You should defer acting on the message's data until your `onPacket` handler
runs. It will only be called when the message's CRC is correct.

Other than that, this library includes no error handling.

This is intentional, as it is optimized for forwarding messages in
resource-constrained environment where most error handling takes too much
time or space.

If you need to discover where in your chain of modules a packet got lost,
the usual process is to include a sequence counter in your packet header.
Increment your packet-loss counter by the number of missed entries in the
sequence, and add a way to retrieve that counter.

Checking whether `onPacket` is *not* called after `onHeader` indicates a
CRC error. That error however can result from various problems (an actual
transmission error, wrong number of bytes written by *some* module before
the current one, etc.). This library doesn't try to determine which is
which.

The master should wait `SP_MAX_FRAME_DELAY` milliseconds between messages,
plus the time required for transmitting the data added by modules.

### Deep Sleep

In some applications, microcontrollers go to "deep sleep" with disabled
oscillators. However, serial communication is difficult when you don't have
a clock.

In this situation it's helpful to set the frame start to `0xFF` and to
hook an interrupt to the rising edge of the serial line (i.e. the end of
the start bit). The interrupt shall disable itself, re-enable the serial
receiver, and call the `wokeUp()` method.

# Usage example

Check [this fork of the diyBMS code](https://github.com/M-o-a-T/diyBMS-code).


debian/changelog
serialpacker (0.10.4-1) unstable; urgency=medium

  * Initial Debianization.

 -- Matthias Urlichs <matthias@urlichs.de>  Fri, 08 Dec 2023 09:52:19 +0100


test.py
"""
Rudimentary serialpacker test
"""
from __future__ import annotations

import sys

from serialpacker import SerialPacker as S


def test():
    """rudimentary test"""
    a = S(max_idle=9999999999, max_packet=300)

    hello = (" ".join(sys.argv[1:]) or "HiThere").encode("utf-8")
    h, data, t = a.frame(hello)
    data = h + data + t
    if len(data) > 15:
        print(len(hello), data[:5], "…", data[-5:])
    else:
        print(len(hello), data)
    for j in range(3):
        for i, b in enumerate(data):
            x = a.feed(b)
            if x is not None and i != len(data) - 1:
                raise RuntimeError("returned %r, pos %d" % (x, i))
        if x is None:
            raise RuntimeError("returned None %d" % (j,))
        if x != hello:
            raise RuntimeError(f"returned {x !r} instead of {hello !r}")


if __name__ == "__main__":
    test()


tests/test_basic.py
"""
Empty test file
"""
from __future__ import annotations

import serialpacker  # noqa:F401 pylint: disable=unused-import


def test_nothing():
    """
    Empty test
    """


serialpacker.py
"""
This module implements a simple framing scheme for serially-transmitted
data. Frames use an optional fixed start byte, a intra-frame time maximum
to protect against data loss, and a CRC16 (polynomial 0xBAAD, optimal for
small-ish frames with a large number of bit errors).

The length is transmitted as one or two characters, depending on its
maximum, in order to keep packets small yet not restrict size too much.

A C implementation of this scheme is also available. The C version supports
streaming; this Python implementation does not.
"""
from __future__ import annotations

__all__ = ["FRAME_START", "SerialPacker", "CRC16"]

try:
    import utime as time
except ImportError:
    # CPython
    import time

    def ticks_ms():
        """System ticks, in msec"""
        return time.clock_gettime_ns(time.CLOCK_MONOTONIC_RAW) / 1000000

    def ticks_diff(a, b):
        """Difference between ticks, non-wrapping"""
        return a - b

else:
    # MicroPython
    ticks_ms = time.ticks_ms
    ticks_diff = time.ticks_diff

FRAME_START = 0x85


class CRC16:
    """
    A simple CRC routine.

    Polynomial: 0xBAAD, no inversion or similar tricks
    """

    # cd moat-bus/python/moatbus
    # python3 ./crc.py -b16 -p0xBAAD -d4
    len = 2

    # Use a nibble-based table; it's fast enough, but doesn't need a
    # kbyte of memory (MicroPython doesn't have arrays).
    table = (
        0x0000,
        0x64A8,
        0xC950,
        0xADF8,
        0xE7FB,
        0x8353,
        0x2EAB,
        0x4A03,
        0xBAAD,
        0xDE05,
        0x73FD,
        0x1755,
        0x5D56,
        0x39FE,
        0x9406,
        0xF0AE,
    )

    def __init__(self):
        self.crc = 0

    def feed(self, byte):
        """add a byte to the CRC"""
        crc = self.crc
        crc = self.table[((byte >> 4) ^ crc) & 0xF] ^ (crc >> 4)
        crc = self.table[(byte ^ crc) & 0xF] ^ (crc >> 4)
        self.crc = crc


class SerialPacker:
    """
    Frame sender and receiver.

    Args:
      max_idle:
        max milliseconds between bytes within a frame.
      max_packet:
        maximum packet size. Size is always transmitted as a single byte if
        max_packet is <256, otherwise two bytes are used whenever the actual
        size is >127.
      frame_start: start byte, or None for solely relying on inter-frame timeout.
      crc:
        CRC method to use. The default is our CRC16 with 0xBAAD polynomial.
    """

    # error counters
    err_crc = 0  # CRC wrong
    err_frame = 0  # length wrong or timeout

    def __init__(
        self,
        max_idle=100,
        max_packet=127,
        frame_start=FRAME_START,
        crc=CRC16,
        mark=None,
    ):
        self.max_packet = max_packet
        self.max_idle = max_idle
        self.frame_start = frame_start
        self.mark = mark
        self.mark_seen = False
        self._crc = crc
        self.reset()

    def reset(self):
        """Reset the packet receiver."""
        self.crc = self._crc()
        self.last = ticks_ms()
        self.buf = None
        self.state = 1 if self.frame_start is None else 0
        self.pos = 0
        # intentionally does not clear nbuf

    def is_idle(self):
        """
        Test whether we're currently receiving a frame.
        """
        if self.s == 0:
            return True
        t = ticks_ms()
        if ticks_diff(t, self.last) < self.max_idle:
            return False
        self.err_frame += 1
        self.reset()
        return True

    def wakeup(self):
        """
        Utility method to start a frame without a start byte. Useful if the
        first byte is used as a wake-from-sleep signal.
        """
        self.reset()
        self.s = 1

    def feed(self, byte):
        """
        Receiver for a single byte.

        Returns: a packet, if this byte completed its reception.
        """
        t = ticks_ms()
        if ticks_diff(t, self.last) >= self.max_idle:
            self.err_frame += 1
            self.reset()
        self.last = t

        s = self.state

        if self.mark is None:
            pass
        elif self.mark_seen:
            self.mark_seen = False
        elif byte == self.mark:
            self.mark_seen = True
            return None
        else:
            return byte

        if s == 0:  # idle
            if self.pos > 0:
                self.pos -= 1
                return byte
            elif byte == self.frame_start:
                s = 1
            else:
                if byte >= 0xC0:
                    # UTF-8 lead-in character. Skip 0x10xxxxxx bytes so
                    # an accidental start byte is not misrecognized.
                    c = byte ^ 0xFF
                    n = 0
                    while c:
                        n += 1
                        c >>= 1
                    self.pos = 6 - n
                return byte

        elif s == 1:  # len1
            if byte == 0:  # zero length = error (break?)
                self.err_frame += 1
                self.reset()
                return None
            self.len = byte
            if self.max_packet > 255 and (byte & 0x80):  # possibly a two byte packet
                s = 2
            elif byte > self.max_packet:
                self.err_frame += 1
                self.reset()
                return None
            else:
                s = 3
                self.buf = bytearray(self.len)
                self.pos = 0

        elif s == 2:  # len2
            self.len = (self.len & 0x7F) | (byte << 7)
            if self.len > self.max_packet:
                self.err_frame += 1
                self.reset()
                return None

            s = 3
            self.buf = bytearray(self.len)
            self.pos = 0

        elif s == 3:
            self.crc.feed(byte)
            self.buf[self.pos] = byte
            self.pos += 1
            if self.pos == self.len:
                s = 4
                self.pos = self.crc.len
                self.crc_in = 0

        elif s == 4:  # CRC
            self.crc_in = (self.crc_in << 8) | byte
            self.pos -= 1
            if self.pos == 0:
                if self.crc.crc != self.crc_in:  # error
                    self.err_crc += 1
                    rbuf = None
                else:
                    rbuf = self.buf
                self.reset()
                return rbuf

        self.state = s
        return None

    def frame(self, data):
        """
        Build framing for a message.

        Returns:
            a triple with head (start byte, length), bytes (possibly
            mark-ed), and tail (CRC) data.
        """
        ld = len(data)
        if ld > self.max_packet:
            raise ValueError(f"max len {self.max_packet}")
        crc = self._crc()
        for b in data:
            crc.feed(b)
        h = bytearray()
        if self.frame_start is not None:
            h.append(self.frame_start)
        if self.max_packet > 255 and ld > 127:
            h.extend(((ld & 0x7F) | 0x80, ld >> 7))
        else:
            h.append(ld)
        t = bytes((crc.crc >> 8, crc.crc & 0xFF))

        if self.mark is not None:

            def itermark(data):
                for b in data:
                    yield self.mark
                    yield b

            h = bytes(itermark(h))
            data = bytes(itermark(data))
            t = bytes(itermark(t))
        return h, data, t


src/SerialPacker.h
// Copyright (c) 2020 Stuart Pittaway
// Copyright (c) 2022 Matthias Urlichs
// https://github.com/M-o-a-T/SerialPacker
//
// GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007

#pragma once

#include <Arduino.h>

//
// Definitions: see README

#ifndef SP_MAX_FRAME_DELAY
#define SP_MAX_FRAME_DELAY 100
#endif

#ifndef SP_MAX_PACKET
#define SP_MAX_PACKET 127
#endif

#ifndef SP_FRAME_START
#define SP_FRAME_START 0x85
#endif

#if SP_MAX_PACKET>255
typedef uint16_t SB_SIZE_T;
#else
typedef uint8_t SB_SIZE_T;
#endif

enum SerialPackerState : uint8_t {
    SP_IDLE=0,
#if SP_FRAME_START >= 0
    SP_LEN1,
#endif
    SP_LEN2,
    SP_DATA0,
    SP_DATA,
    SP_CRC1,
    SP_CRC2,
    SP_DONE, // wait for getting data
    SP_ERROR, // wait for idle
};

class SerialPacker
{
public:
    SerialPacker() {}
    typedef void (*PacketHandlerFunction)();

    void begin(Stream *_stream, PacketHandlerFunction onHeader, PacketHandlerFunction onReader, PacketHandlerFunction onPacket, uint8_t *receiveBuf, SB_SIZE_T bufSize, uint8_t headerSize=0)
    {
        stream = _stream;
        onHeaderReceived = onHeader;
        onReadReceived = onReader;
        onPacketReceived = onPacket;
        receiveBuffer = receiveBuf;
        receiveBufferLen = bufSize;
        headerLen = headerSize;
#ifdef SP_ERRCOUNT
        clearErrors();
#endif
    }

    void checkInputStream();
    bool isIdle() {
        return receiveState == SP_IDLE;
    }

#if SP_FRAME_START >= 0
    void wokeUp() {
        if(isIdle())
            receiveState = SP_LEN1;
    }
#endif

    // start sending
    void sendStartFrame(SB_SIZE_T length);
    void sendStartCopy(SB_SIZE_T addLength);
    void sendDefer(SB_SIZE_T readLength)
    {
        readLen += readLength;
    }

    void sendBuffer(const void *buffer, SB_SIZE_T length);
    void sendByte(uint8_t data);

    // stop sending
    void sendEndFrame(bool broken=false);

/*    
    void debugByte(uint8_t data)
    {
        if (data <= 0x0F)
        {
            Serial1.write('0');
        }
        Serial1.print(data, HEX);
        Serial1.write(' ');
    }
*/

    bool isCopying() {
        return copyInput;
    }

    SB_SIZE_T receiveCount() {
        return receivePos;
    }

#ifdef SP_ERRCOUNT
    uint16_t errCRC = 0;
    uint16_t errTimeout = 0;

    inline void clearErrors() {
        errCRC = 0;
        errTimeout = 0;
    }
#endif

    static uint16_t crc16_update(uint16_t crc, uint8_t byte);
    static uint16_t crc16_buffer(uint8_t data[], uint16_t length);
private:

    // receiver *****

    SerialPackerState receiveState = SP_IDLE;
#ifdef SP_MARK
    bool receiveMark = false;
#endif

    uint16_t receiveCRC;

    uint16_t last_ts = 0;

    SB_SIZE_T headerLen = 0;
    SB_SIZE_T readLen = 0;
    bool copyInput = false;
    uint8_t crcHi;

    //Pointer to start of receive buffer (byte array)
    uint8_t *receiveBuffer = nullptr;
    //Index into receive buffer of current position
    SB_SIZE_T receivePos = 0;
    SB_SIZE_T receiveLen = 0;
    SB_SIZE_T receiveBufferLen = 0;

    //Send/Receive stream
    Stream *stream = nullptr;

    // process an incoming byte
    void processByte(uint8_t data);

    //Call back: headerSize bytes have been received
    PacketHandlerFunction onHeaderReceived = nullptr;
    //Call back: readLen bytes have been received
    PacketHandlerFunction onReadReceived = nullptr;
    //Call back: a complete message has been received
    PacketHandlerFunction onPacketReceived = nullptr;

    // sender *****

    uint16_t sendCRC;
#ifdef SP_SENDLEN
    SB_SIZE_T sendPos = 0;
    SB_SIZE_T sendLen = 0;
#endif

    void reset()
    {
        receiveState = SP_IDLE;
    }
};


serialpacker/__init__.py
"""
This module implements a simple framing scheme for serially-transmitted
data. Frames use an optional fixed start byte, a intra-frame time maximum
to protect against data loss, and a CRC16 (polynomial 0xBAAD, optimal for
small-ish frames with a large number of bit errors).

The length is transmitted as one or two characters, depending on its
maximum, in order to keep packets small yet not restrict size too much.

A C implementation of this scheme is also available. The C version supports
streaming; this Python implementation does not.
"""

__all__ = ["FRAME_START", "SerialPacker", "CRC16"]

try:
    import utime as time
except ImportError:
    # CPython
    import time

    def ticks_ms():
        """System ticks, in msec"""
        return time.clock_gettime_ns(time.CLOCK_MONOTONIC_RAW) / 1000000

    def ticks_diff(a, b):
        """Difference between ticks, non-wrapping"""
        return a - b

else:
    # MicroPython
    ticks_ms = time.ticks_ms
    ticks_diff = time.ticks_diff

FRAME_START = 0x85


class CRC16:
    """
    A simple CRC routine.

    Polynomial: 0xBAAD, no inversion or similar tricks
    """

    # cd moat-bus/python/moatbus
    # python3 ./crc.py -b16 -p0xBAAD -d4
    len = 2

    # Use a nibble-based table; it's fast enough, but doesn't need a
    # kbyte of memory (MicroPython doesn't have arrays).
    table = [
        0x0000,
        0x64A8,
        0xC950,
        0xADF8,
        0xE7FB,
        0x8353,
        0x2EAB,
        0x4A03,
        0xBAAD,
        0xDE05,
        0x73FD,
        0x1755,
        0x5D56,
        0x39FE,
        0x9406,
        0xF0AE,
    ]

    def __init__(self):
        self.crc = 0

    def feed(self, byte):
        """add a byte to the CRC"""
        crc = self.crc
        crc = self.table[((byte >> 4) ^ crc) & 0xF] ^ (crc >> 4)
        crc = self.table[(byte ^ crc) & 0xF] ^ (crc >> 4)
        self.crc = crc


class SerialPacker:
    """
    Frame sender and receiver.

    Args:
      max_idle:
        max milliseconds between bytes within a frame.
      max_packet:
        maximum packet size. Size is always transmitted as a single byte if
        max_packet is <256, otherwise two bytes are used whenever the actual
        size is >127.
      frame_start: start byte, or None for solely relying on inter-frame timeout.
      crc:
        CRC method to use. The default is our CRC16 with 0xBAAD polynomial.
    """

    # error counters
    err_crc = 0  # CRC wrong
    err_frame = 0  # length wrong or timeout

    def __init__(
        self, max_idle=100, max_packet=127, frame_start=FRAME_START, crc=CRC16, mark=None
    ):
        self.max_packet = max_packet
        self.max_idle = max_idle
        self.frame_start = frame_start
        self.mark = mark
        self.mark_seen = False
        self._crc = crc
        self.reset()

    def reset(self):
        """Reset the packet receiver."""
        self.crc = self._crc()
        self.last = ticks_ms()
        self.buf = None
        self.state = 1 if self.frame_start is None else 0
        self.pos = 0
        # intentionally does not clear nbuf

    def is_idle(self):
        """
        Test whether we're currently receiving a frame.
        """
        if self.s == 0:
            return True
        t = ticks_ms()
        if ticks_diff(t, self.last) < self.max_idle:
            return False
        self.err_frame += 1
        self.reset()
        return True

    def wakeup(self):
        """
        Utility method to start a frame without a start byte. Useful if the
        first byte is used as a wake-from-sleep signal.
        """
        self.reset()
        self.s = 1

    def feed(self, byte):
        """
        Receiver for a single byte.

        Returns: a packet, if this byte completed its reception.
        """
        t = ticks_ms()
        if ticks_diff(t, self.last) >= self.max_idle:
            self.err_frame += 1
            self.reset()
        self.last = t

        s = self.state

        if self.mark is None:
            pass
        elif self.mark_seen:
            self.mark_seen = False
        elif byte == self.mark:
            self.mark_seen = True
            return None
        else:
            return byte

        if s == 0:  # idle
            if self.pos > 0:
                self.pos -= 1
                return byte
            elif byte == self.frame_start:
                s = 1
            else:
                if byte >= 0xC0:
                    # UTF-8 lead-in character. Skip 0x10xxxxxx bytes so
                    # an accidental start byte is not misrecognized.
                    c = byte ^ 0xFF
                    n = 0
                    while c:
                        n += 1
                        c >>= 1
                    self.pos = 6 - n
                return byte

        elif s == 1:  # len1
            if byte == 0:  # zero length = error (break?)
                self.err_frame += 1
                self.reset()
                return None
            self.len = byte
            if self.max_packet > 255 and (byte & 0x80):  # possibly a two byte packet
                s = 2
            elif byte > self.max_packet:
                self.err_frame += 1
                self.reset()
                return None
            else:
                s = 3
                self.buf = bytearray(self.len)
                self.pos = 0

        elif s == 2:  # len2
            self.len = (self.len & 0x7F) | (byte << 7)
            if self.len > self.max_packet:
                self.err_frame += 1
                self.reset()
                return None

            s = 3
            self.buf = bytearray(self.len)
            self.pos = 0

        elif s == 3:
            self.crc.feed(byte)
            self.buf[self.pos] = byte
            self.pos += 1
            if self.pos == self.len:
                s = 4
                self.pos = self.crc.len
                self.crc_in = 0

        elif s == 4:  # CRC
            self.crc_in = (self.crc_in << 8) | byte
            self.pos -= 1
            if self.pos == 0:
                if self.crc.crc != self.crc_in:  # error
                    self.err_crc += 1
                    rbuf = None
                else:
                    rbuf = self.buf
                self.reset()
                return rbuf

        self.state = s
        return None

    def frame(self, data):
        """
        Build framing for a message.

        Returns:
            a triple with head (start byte, length), bytes (possibly
            mark-ed), and tail (CRC) data.
        """
        ld = len(data)
        if ld > self.max_packet:
            raise ValueError(f"max len {self.max_packet}")
        crc = self._crc()
        for b in data:
            crc.feed(b)
        h = bytearray()
        if self.frame_start is not None:
            h.append(self.frame_start)
        if self.max_packet > 255 and ld > 127:
            h.extend(((ld & 0x7F) | 0x80, ld >> 7))
        else:
            h.append(ld)
        t = bytes((crc.crc >> 8, crc.crc & 0xFF))

        if self.mark is not None:

            def itermark(data):
                for b in data:
                    yield self.mark
                    yield b

            h = bytes(itermark(h))
            data = bytes(itermark(data))
            t = bytes(itermark(t))
        return h, data, t


src/SerialPacker.cpp
// Copyright (c) 2020 Stuart Pittaway
// Copyright (c) 2022 Matthias Urlichs
// https://github.com/M-o-a-T/SerialPacker
//
// GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007

#include "SerialPacker.h"

#ifndef SP_CRC
#define SP_CRC 0xBAAD
#endif

// generated by:
// git clone https://github.com/M-o-a-T/moat-bus
// cd moat-bus/python/moatbus
// python3 ./crc.py -b16 -p0xBAAD -d4
//
static const uint16_t crc16_4[] = {
#if SP_CRC==0xBAAD
0x0000, 0x64a8, 0xc950, 0xadf8,
0xe7fb, 0x8353, 0x2eab, 0x4a03,
0xbaad, 0xde05, 0x73fd, 0x1755,
0x5d56, 0x39fe, 0x9406, 0xf0ae,
#else
#error You need to generate a CRC table for this polynomial
#endif
};

uint16_t SerialPacker::crc16_update(uint16_t crc, uint8_t byte)
{
    crc = crc16_4[((byte>>4) ^ crc) & 0xF] ^ (crc>>4);
    crc = crc16_4[(byte ^ crc) & 0xF] ^ (crc>>4);
    return crc;
}

uint16_t SerialPacker::crc16_buffer(uint8_t data[], uint16_t length)      
{ 
    uint16_t crc = 0;
    while(length--)
        crc = crc16_update(crc, *(data++));
    return crc;
}

void SerialPacker::processByte(uint8_t data)
{
    //debugByte(data);

    uint16_t ts = millis(); // ignores the high word
    if(ts-last_ts > SP_MAX_FRAME_DELAY && receiveState != SP_IDLE)
    {
#ifdef SP_TRACE
        SP_TRACE.println(" R");
#endif
#ifdef SP_MARK
        receiveMark = false;
#endif
        receiveState = SP_IDLE;
#ifdef SP_ERRCOUNT
        errTimeout += 1;
#endif
    }
    last_ts = ts;
#ifdef SP_MARK
    if (receiveMark)
        receiveMark = false;
    else if (data == SP_MARK)
        receiveMark = true;
    else {
#ifdef SP_NONFRAME_STREAM
        SP_NONFRAME_STREAM.write(data);
#else
        stream->write(data);
#endif
    }
#endif

    switch(receiveState) {
    case SP_IDLE:
#if SP_FRAME_START >= 0
        if(data != SP_FRAME_START) {
#ifdef SP_NONFRAME_STREAM
#ifdef SP_TRACE
            SP_TRACE.write('?');
            SP_TRACE.print((int)data);
            SP_TRACE.write(' ');
#endif
            SP_NONFRAME_STREAM.write(data);
#endif
            break;
        }
        receiveState = SP_LEN1;
        break;
    case SP_LEN1:
#endif
        if (data == 0x00) {
            // A zero-length packet doesn't make sense; it's most likely a
            // BREAK.
            reset();
            break;
        }
#if SP_MAX_PACKET>255
        receiveLen = data&0x7F;
        if(data & 0x80)
            receiveState = SP_LEN2;
        else
#else
        receiveLen = data;
#endif
            // the next line is conditional on data & 0x80 if SP_MAX_PACKET>255
            receiveState = SP_DATA0;
        receivePos = 0;
        readLen = 0;
        break;
    case SP_LEN2:
        receiveLen |= data<<7;
        receiveState = SP_DATA0;
        break;
    case SP_DATA0:
        if(receiveLen > SP_MAX_PACKET) {
            receiveState = SP_ERROR;
            break;
        }
        receiveCRC = 0;
        receivePos = 0;
        if(receiveLen == 0) {
            receiveState = SP_CRC1;
            if(headerLen == 0 && onHeaderReceived)
                (*onHeaderReceived)();
            break;
        }
        receiveState = SP_DATA;
        // fall thru
    case SP_DATA:
        receiveCRC = crc16_update(receiveCRC, data);
        if(receivePos < receiveBufferLen)
            receiveBuffer[receivePos] = data;
        receivePos += 1;
        if (readLen) {
            readLen--;
            if (readLen == 0) {
                if (onReadReceived != nullptr)
                    (*onReadReceived)();
            }
        }
        else if(copyInput) {
            sendByte(data);
        }
        if(receivePos == receiveLen)
            receiveState = SP_CRC1;
        if(receivePos == headerLen) {
            if (onHeaderReceived != nullptr)
                (*onHeaderReceived)();
        }
        break;
    case SP_CRC1:
        crcHi = data;
        receiveState = SP_CRC2;
        break;
    case SP_CRC2:
        {
            uint16_t crc = ((crcHi<<8) | data);
#ifdef SP_TRACE
            SP_TRACE.print(crc,HEX);
            SP_TRACE.write(' ');
#endif
            if(crc ^ receiveCRC) {
#ifdef SP_TRACE
                SP_TRACE.print(F("-CRC:rem "));
                SP_TRACE.print(crc, HEX);
                SP_TRACE.print(F(" loc "));
                SP_TRACE.println(receiveCRC, HEX);
#endif
#ifdef SP_ERRCOUNT
                if (crc != 1)
                    errCRC += 1;
#endif
                receiveState = SP_IDLE;
                if(copyInput)
                    sendEndFrame(true);
            } else {
#ifdef SP_TRACE
                SP_TRACE.println("+CRC");
#endif
                receiveLen = receivePos;
                if(onPacketReceived != nullptr) {
                    (*onPacketReceived)();
                    if(copyInput)
                        sendEndFrame(false);
                }
                receiveState = SP_IDLE;
            }
        }
        break;
    case SP_DONE:
        break;
    case SP_ERROR:
        break;
    }
}

// Checks stream for new bytes to arrive and processes them as needed
void SerialPacker::checkInputStream()
{
    if (stream == nullptr || receiveBuffer == nullptr)
        return;

    while (stream->available() > 0) {
        uint16_t data = stream->read();
#ifdef SP_TRACE
        SP_TRACE.print(data,HEX);
        SP_TRACE.write(' ');
#endif
        processByte((uint8_t)data);
    }
}

void SerialPacker::sendByte(uint8_t data)
{
    //debugByte(data);
#ifdef SP_SENDLEN
    if(sendPos++ > sendLen)
        return; // oops
#endif
    sendCRC = crc16_update(sendCRC, data);
#ifdef SP_MARK
    stream->write(SP_MARK);
#endif
    stream->write(data);
}

// Sends a buffer (fixed length)
void SerialPacker::sendBuffer(const void *buffer, SB_SIZE_T len)
{
    const uint8_t *buf = (const uint8_t *)buffer;
    for (SB_SIZE_T i = 0; i < len; i++,buf++)
        sendByte(*buf);
}

void SerialPacker::sendStartFrame(SB_SIZE_T length)
{
#if SP_FRAME_START >= 0
#ifdef SP_MARK
    stream->write(SP_MARK);
#endif
    stream->write(SP_FRAME_START);
#endif
#if SP_MAX_PACKET>255
    if(length > 0x7F) {
#ifdef SP_MARK
        stream->write(SP_MARK);
#endif
        stream->write(length | 0x80);
#ifdef SP_MARK
        stream->write(SP_MARK);
#endif
        stream->write(length >> 7);
    } else
#endif
#ifdef SP_MARK
    stream->write(SP_MARK);
#endif
    stream->write(length);
    sendCRC = 0;
#ifdef SP_SENDLEN
    sendPos = 0;
    sendLen = length;
#endif
    copyInput = false;
}

void SerialPacker::sendStartCopy(SB_SIZE_T addLength)
{
    // ASSERT(!copyInput);

    // Subtract any skipped bytes (i.e.. those read beyond @headerlen) from
    // the resulting header. The receive buffer is guaranteed to contain
    // @headerLen bytes at this point.
    SerialPacker::sendStartFrame(receiveLen + addLength - (receivePos-headerLen));

    for(uint8_t i = 0; i < headerLen; i++)
        sendByte(receiveBuffer[i]);
    copyInput = true;
}

void SerialPacker::sendEndFrame(bool broken)
{
#ifdef SP_TRACE
    SP_TRACE.write(broken ? '-' : '+'),
    SP_TRACE.println(F("SEND"));
#endif
    uint8_t br = broken;
    copyInput = false;
#ifdef SP_SENDLEN
    if(sendPos != sendLen) {
        // oops
        br = 1;
        while(sendPos < sendLen)
            sendByte(0);
    }
#endif

    uint16_t crc = sendCRC;
    if(br)
        crc ^= 1;
#ifdef SP_MARK
    stream->write(SP_MARK);
#endif
    stream->write(crc >> 8);
#ifdef SP_MARK
    stream->write(SP_MARK);
#endif
    stream->write(crc & 0xFF);

#ifdef SP_TRACE
    SP_TRACE.print(F("\nX"));
#ifdef SP_SENDLEN
    SP_TRACE.print(sendLen);
    if(br>1) {
        SP_TRACE.print(F("-BR:"));
        SP_TRACE.println(br-1);
    }
    if(sendPos>sendLen) {
        SP_TRACE.print(F("-POS:"));
        SP_TRACE.print(sendLen);
        SP_TRACE.write(':');
        SP_TRACE.println(sendPos);
    }
#endif // SENDLEN
    if(br) {
        SP_TRACE.print(F("_ERR:"));
        SP_TRACE.println(br);
    } else
        SP_TRACE.println("+");
#endif // TRACE
}

